### DISCUSSION TITLE
Integration testing backend widgets

### DISCUSSION THREAD

**Discussion by ericp-mrel** [wintercms/winter#186] [Q&A] [answered]
Is there any sort of documentation which explains how to perform integration testing with phpunit on backend widgets?
I'm trying to test my custom backend widget which supports file uploads, but I'm not sure what the cleanest/best way is to go about it.

Here's my current test code, and for some reason this test works locally but when it runs in github actions it fails with an HTTP error code of 403.
`Expected status code 200 but received 403. Failed asserting that false is true.`

```php
    public function test_it_can_upload_to_an_uploadable_widget()
    {
        $this->app->singleton('backend.auth', function ($app) {
            $app['auth.loaded'] = true;

            return AuthManager::instance();
        });

        $user = new BackendUser();
        $this->actingAs($user);

        $controller = new Controller();

        $widget = new myCustomWidget($controller, 'uploader');
        $widget->bindToController();

        $this->app['router']->post('/upload', function () use ($controller) {
            return $controller->run();
        });

        $fileUpload = new UploadedFile(
            $this->getTestFilesDirectory('test.jpg'),
            'alternativename.jpg',
            'image/jpeg',
            filesize($this->getTestFilesDirectory('test.jpg'))
        );

        $result = $this->post('upload', ['file_data' => $fileUpload], [
            'X-Requested-With' => 'XMLHttpRequest',
            'X-WINTER-REQUEST-HANDLER' => $widget->getEventHandler('onUpload'),
        ]);

        $result
            ->assertStatus(200)
            ->assertJson([
                'result' => 'success',
            ]);
    }
```

Any help is appreciated, Thanks!

**Comment by LukeTowers** [comment]
403 is an authentication error, so perhaps you're having some problems with the actingAs() method in github actions?

**Reply by ericp-mrel** [reply]
Yeah, it's very weird because I basically took the "backend login" part of the test from here: https://github.com/wintercms/winter/blob/develop/tests/unit/backend/widgets/FormTest.php#L42-L43
Unless it has to do with how I am creating a new backend controller instance? But still weird that it works just fine locally, just not in Github.

Not sure why it would work when your tests are run from github actions, but not mine.

**Reply by ericp-mrel** [reply]
Based on some of the different things I've tried out, it seems like the issue is coming from the backend controller class throwing the 403, because when I do the following check it passes: `$this->assertTrue(AuthManager::instance()->check());` Which means that the auth manager does at least have my current user object, but it seems like the controller instance isn't getting one?

But again, I'm not sure if there's a more correct way to instantiate a backend controller for testing purposes.

**Comment by bennothommo** [comment]
@ericp-mrel Check out the [Dusk plugin](https://github.com/wintercms/wn-dusk-plugin/) and the tests within - it might give you some pointers on how to correctly authenticate and test front-end functionality.

**Reply by ericp-mrel** [reply]
I have looked at this plugin's tests, but I'm not really looking at doing full front-end/browser testing. I just want to perform some tests with phpunit to test my widgets file upload functionality.

**Comment by ericp-mrel** [comment, marked as answer]
Ahh, I figured out my problem. I was copying the cms.php file from `config/cms.php` into `config/testing/cms.php` because I saw the Winter CMS `tests.yml` file doing that, but I didn't think anything of it until I realized by doing that I was effectively enabling CSRF protection which was the reason I was seeing the 403 error. So, not overriding the supplied cms.php in the testing config folder fixed my issue so that CSRF protection is not being enable during testing.

Glad I've finally figured out my issue and that explains why it was happening in Github Actions and not my local copy.
